src/persist/README.md

```diff
@@ -1,51 +1,359 @@
+    collection" jargon for the equivalent concept at the level of the platform
+    architecture's [storage layer] (design doc in progress), and so call this
+    something else.
+  - _definite_: Quoting the [correctness doc], "A collection is definite when
+    all uses of any one identification will yield exactly the same time-varying
+    collection of data (identical contents at each logical time). A handle to a
+    collection can be "definite from since until upper" (names pending) if it
+    reflects the definite results for times greater or equal to since but not
+    greater or equal to upper; this intentionally allows for compaction of
+    historical detail and unavailability of future data, without ambiguity about
+    which logical times can be described with certainty."
+  - _(external) collection name_: A human-readable, process-unique `&str`
+    identifier for a persisted collection. Internally, this name gets
+    transparently mapped 1:1 to a smaller collection id.
+  - _(internal) collection id_: The internal equivalent of a collection name.
+    TODO: For efficiency, this is currently exposed in the public API in one
+    place, but it'd be nice to eliminate this.
+  - _write_: Durably adding a set of records to a persisted collection.
+  - _seal_: Durably advancing the frontier of a persisted collection. NB: A
+    common initial misconception when encountering persist is that _write_ is
+    like a filesystem write with minimal durability guarantees and _seal_ like
+    is like fsync, but this incorrect. A write whose future has resolved as
+    successful is durable, regardless of when or if it's sealed.
+  - _storage_: An external source of data durability. All of persist's
+    interactions with the outside world are abstracted through the [Blob] and
+    [Log] traits, together collectively called storage. One valid way to think
+    of persist is as a layer that maps simpler durable storage semantics to ones
+    that are more native to differential dataflow.
+  - _blob_: A durable `&str` key -> (possibly large) `&[u8]` value store. The
+    production implementation is S3 (or other cloud storage). File and in-mem
+    HashMap implementations exist for testing.
+  - _log_: A durable log with `&[u8]` entries. Currently unused, but eventually
+    this will be an optional latency optimization. No production implementations
+    exist yet.
+  - _storage location_: A logical root of data storage. Examples include an s3
+    bucket + prefix or a filesystem folder.
+  - _runtime_: An instantiation of persist. Accessed externally via a
+    thread-safe and clone-able "client".
+
+[OrdKeySpine]: differential_dataflow::trace::implementations::ord::OrdKeySpine
+[Blob]: crate::storage::Blob
+[Log]: crate::storage::Log
+[storage layer]: #integration-with-platform
+
+
+# Rust API
+
+Persist has 2 public APIs: a rust one and a set of differential dataflow
```

Comment:
So I always get a little confused when I try to actually _find_ these APIs. Is something like this right?

> The raw Rust API lives in the `persist::indexed` module, while the differential dataflow operators live in `persist::operators`

Also "Rust API" is a bit ambiguous because the operators are also a Rust API. Low level? Basic? Timely-free? I'm sure there's a good clarifying name somewhere.

Reply:
I think adding the modules here is a great idea! I'll do that.

Atm the "Rust API" is in `persist::indexed::runtime`, but the runtime module is going to get broken up sometime soon (as soon as we have few enough PRs in flight that the refactor won't be too disruptive. After the refactor, it'll probably be something like `persist::client`.

> Also "Rust API" is a bit ambiguous because the operators are also a Rust API. Low level? Basic? Timely-free? I'm sure there's a good clarifying name somewhere.

Hmm, the Rust API stuff all returns Futures (or should, there's a couple that are currently blocking but that's very fixable). How about Async API and Operator API?